### ISSUE TITLE
Two-Factor authentication with SMS

### ISSUE THREAD

**Issue by essivision** [tymondesigns/jwt-auth#917]
Hi,
I want to authenticate with sms verification. As matter of fact I want to implement 2-factor authentication like gmail with jwt-auth library. What should I do? What's your idea?

Thanks

**Comment by zeshan-inquiron** [comment]
@essivision Did you find something for this? I am also looking for 2FA with JWT.

Thanks

**Comment by essivision** [comment]
No I don't find any useful things. but I think it doesn't very hard to write manually.

**Comment by mithek** [comment]
Authenticate the user with the laravel Auth:attempt. If true sent sms. 
The second api call will validate the code from the sms if true then generate the JWT token from user model. 

Api call 1: email + password
Api call 2: email + sms code

**Comment by 3lpsy** [comment]
I would like some advice on this approach.

Create two custom extra claims: mfr for multi-factor required, and mfs for multi-factor satisfied. When a user logins in with password, within the token, add the mfr required claim as either 'sms', 'email', 'totp', or 'authy', an array of these values if requireing more than 2, or false/empty array if mfa is disabled. Then add an mfs claim as false or empty array if more than two.

If mfs does not equal mfr, the user needs to be redirected to the correct mfr page (i.e #/auth/mfa/sms or #/auth/mfa/authy).

So create two middlewares, one for is password auth (normal), and is multi factor auth (mfa). If mfa fails, return the custom requires mfa respose and handle in client. Axios provides response interceptors so you could return 403 with a special header or value and catch it there. 

Basically, this approach implements the extra_claims feature of jwt. It is also absolutely critical to  to check that mfr matches mfs after validating the token as well checking the signature so a user can't edit the mfs value (though the signature check is always critical for sub and done by tymon).

**Comment by ghost** [comment]
@3lpsy did you by any chance was able to do a 2FA with JWT? I'm looking to implement this kind of authentication as well

**Comment by ghost** [comment]
Hi @3lpsy @johnrobertsonsquid I'm looking into this too - did you get anywhere?
